### PR TITLE
testcase/eventloop: Add utc of eventloop

### PR DESCRIPTION
**Pull request by tyha1115**
Add utc of eventloop APIs.
(eventloop_loop APIs, eventloop_timer APIs)

### CODE REVIEW

File: apps/examples/testcase/ta_tc/eventloop/utc/Kconfig [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# For a description of the syntax of this configuration file,
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config EXAMPLES_TESTCASE_EVENTLOOP_UTC
+	bool "Eventloop UTC TestCase Example"
+	default n
+	depends on EVENTLOOP
+	---help---
+		Enable the Task Manager TestCase
```
Comment (jeongchanKim) [inline apps/examples/testcase/ta_tc/eventloop/utc/Kconfig]: Please check this description.

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+	timer_data = malloc(sizeof(strlen(EL_TIMER_DATA)));
+	timer_repeat_data = malloc(sizeof(strlen(EL_TIMER_REPEAT_DATA)));
+	strncpy(timer_data, EL_TIMER_DATA, strlen(EL_TIMER_DATA) + 1);
+	strncpy(timer_repeat_data, EL_TIMER_REPEAT_DATA, strlen(EL_TIMER_REPEAT_DATA) + 1);
+	
+	el_timer_flag = 0;
+	timer = eventloop_add_timer(0, 0, el_timer_cb, timer_data);
+	(void)eventloop_loop_run();
+	TC_ASSERT_EQ_CLEANUP("eventloop_add_timer", el_timer_flag, 1, free(timer_data));
```
Comment (jeongchanKim) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Is it okay not to free timer_repeat_data?

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+	timer = eventloop_add_timer_async(0, 0, el_timer_cb, timer_data);
+	sleep_cnt = 0;
+	while (sleep_cnt < 5) {
+		sleep(1);
+		sleep_cnt++;
+		if (el_timer_flag == 1) {
+			break;
+		}
+	}
+	TC_ASSERT_EQ("eventloop_add_timer_async", el_timer_flag, 1);
```
Comment (jeongchanKim) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Please check free for allocated variable

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+static void utc_eventloop_delete_timer_p(void)
+{
+	el_timer_t *del_timer;
+	int ret;
+	
+	el_timer_flag = 0;
+	del_timer = eventloop_add_timer_async(0, 0, el_delete_timer_cb, NULL);
+	ret = eventloop_delete_timer(del_timer);
```
Comment (jeongchanKim) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: I think you'd better check del_timer is okay or NULL before passing to eventloop_delete_timer().

**Comment by seinfra** [comment]
Target : [0d87141a88dc80475ee8d758cfc864aacf604a12] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [0d87141a88dc80475ee8d758cfc864aacf604a12] - Code Rule Check (C++) OK.

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+{
+	int ret;
+
+	ret = eventloop_loop_run();
+	TC_ASSERT_EQ("eventloop_loop_run", ret, OK);
+
+	TC_SUCCESS_RESULT();
+}
+
+static void utc_eventloop_loop_stop_n(void)
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Why do you add empty function?

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	timer_data = malloc(sizeof(strlen(EL_TIMER_DATA)));
+	timer_repeat_data = malloc(sizeof(strlen(EL_TIMER_REPEAT_DATA)));
+	strncpy(timer_data, EL_TIMER_DATA, strlen(EL_TIMER_DATA) + 1);
+	strncpy(timer_repeat_data, EL_TIMER_REPEAT_DATA, strlen(EL_TIMER_REPEAT_DATA) + 1);
+
+	el_timer_flag = 0;
+	timer = eventloop_add_timer_async(0, 0, el_timer_cb, timer_data);
+	sleep_cnt = 0;
+	while (sleep_cnt < 5) {
+		sleep(1);
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: include <unistd.h>

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+	TC_SUCCESS_RESULT();
+}
+
+static void utc_eventloop_add_timer_p(void)
+{
+	char *timer_data;
+	char *timer_repeat_data;
+
+	timer_data = malloc(sizeof(strlen(EL_TIMER_DATA)));
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Need strlen + 1 for null

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c (hunk 3 above)
Comment (tyha1115) [reply apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Done

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+	printf("Timer callback is called!\n");
+	if (strncmp((char *)data, EL_TIMER_DATA, strlen(EL_TIMER_DATA) + 1) == 0) {
+		el_timer_flag++;
+	}
+	return EVENTLOOP_CALLBACK_CONTINUE;
+}
+
+static bool repeat_timer_cb(void *data)
+{
+	printf("Timer repeat callback is called!\n");
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Needs stdio.h

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+static bool loop_stop_flag;
+
+static int send_cnt;
+
+static el_timer_t *g_repeat_timer;
+
+static el_event_t *wifi_on_handle;
+static el_event_t *wifi_off_handle;
+
+static bool timer_cb(void *data)
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Needs stdbool.h

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+{
+	int sleep_cnt;
+	el_timer_t *timer;
+	el_timer_t *repeat_timer;
+
+	el_timer_flag = 0;
+	timer = eventloop_add_timer_async(0, 0, timer_cb, EL_TIMER_DATA);
+	sleep_cnt = 0;
+	while (sleep_cnt < 5) {
+		sleep(1);
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: needs header

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+    el_thread_safe_flag = false;
+    ret = eventloop_thread_safe_function_call(el_thread_safe_cb, EL_THREAD_SAFE_DATA);
+    eventloop_loop_run();
+    TC_ASSERT_EQ("eventloop_thread_safe_function_call", ret, OK);
+    TC_ASSERT_EQ("eventloop_thread_safe_function_call", el_thread_safe_flag, true);
+
+    TC_SUCCESS_RESULT();
+}
+
+#ifdef CONFIG_BUILD_KERNEL
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Needs header

**Comment by seinfra** [comment]
Target : [19daf134feb7341618cb8e86fa6322961e58b23c] - Code Rule Check Result:
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:330: ERROR: [SPC_M_SEP] space prohibited between function name and open parenthesis '('
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:353: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:354: ERROR: [IDT_M_TAB] code indent should use tabs where possible
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:354: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:355: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:360: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:361: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:362: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:363: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:365: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:370: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:372: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:373: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:374: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:375: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:376: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:378: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [19daf134feb7341618cb8e86fa6322961e58b23c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3c25a5f060ea91d36c70dad5f7b4dbb79d714374] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3c25a5f060ea91d36c70dad5f7b4dbb79d714374] - Code Rule Check Result:
apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c:331: ERROR: [SPC_M_SEP] space prohibited between function name and open parenthesis '('

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c (hunk 8 above)
Comment (tyha1115) [reply apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: `stdio.h` is added.

**Comment by seinfra** [comment]
Target : [305224f428d35d87abc7ba32cd14da403c8b6d95] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [305224f428d35d87abc7ba32cd14da403c8b6d95] - Code Rule Check OK.

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+		wifi_off_handle = NULL;
+		return EVENTLOOP_CALLBACK_STOP;
+	}
+	return EVENTLOOP_CALLBACK_CONTINUE;
+}
+
+static void utc_eventloop_send_event_p(void)
+{
+	send_cnt = 0;
+	(void)eventloop_add_timer(2000, true, (timeout_callback)eventloop_send_event_func, NULL);
```
Comment (jeongarmy) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Checking the return value of eventloop_send_event is missed

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 13]
```diff
@@ -0,0 +1,420 @@
...
+	TC_SUCCESS_RESULT();
+}
+
+static int wifi_on_callback(void *data, void *event_data)
+{
+	static int cnt = 0;
+
+	printf("WIFI ON CALLBACK!! Callback data: %s\n", (char *)data);
+	if (event_data != NULL) {
+		printf("Event data : %s\n", (char *)event_data);
```
Comment (jeongarmy) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Please check wheter this event_data is same with data which sent from eventloop_send_event

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 14]
```diff
@@ -0,0 +1,420 @@
...
+}
+
+static void utc_eventloop_add_timer_async_p(void)
+{
+	int sleep_cnt;
+	el_timer_t *timer;
+	el_timer_t *repeat_timer;
+
+	el_timer_flag = 0;
+	timer = eventloop_add_timer_async(0, 0, (timeout_callback)timer_cb, EL_TIMER_DATA);
```
Comment (jeongarmy) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: change 0 to false

Comment (jeongarmy) [reply apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: interval 0 is right?

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+
+static void utc_eventloop_delete_timer_n(void)
+{
+	int ret;
+
+	ret = eventloop_delete_timer(NULL);
+	TC_ASSERT_EQ("eventloop_delete_timer", ret, EVENTLOOP_INVALID_PARAM);
+
+	ret = eventloop_delete_timer(g_repeat_timer);
+	TC_ASSERT_EQ("eventloop_delete_timer", ret, EVENTLOOP_INVALID_HANDLE);
```
Comment (jeongarmy) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: This is same case with line 192.
please move and add comment for this.

**Comment by seinfra** [comment]
Target : [a9360d394626fb9fb093205b53ac863dc6c48d7d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [a9360d394626fb9fb093205b53ac863dc6c48d7d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c1987c937f0d35c747bfec46a8533af43889391d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [c1987c937f0d35c747bfec46a8533af43889391d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [214216b7d068e04192ffb8e15fe217663efdf87a] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [214216b7d068e04192ffb8e15fe217663efdf87a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [10a99efd247370dc45e57c04d23c3f0bd0758665] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [10a99efd247370dc45e57c04d23c3f0bd0758665] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [eacf683ee086004d76f8a7ea6ac2fe3aecab4f87] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [eacf683ee086004d76f8a7ea6ac2fe3aecab4f87] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [5453ffb96ddabba420a1de76ccd395aaefab7a9d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [5453ffb96ddabba420a1de76ccd395aaefab7a9d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ecc5a8de608f422c44a9ae5664d30201153c63d6] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ecc5a8de608f422c44a9ae5664d30201153c63d6] -

**Comment by seinfra** [comment]
Target : [fbe372ec8ab15f3ec78d40379e9f7c858172ef9c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fbe372ec8ab15f3ec78d40379e9f7c858172ef9c] - Code Rule Check OK.

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 16]
```diff
@@ -0,0 +1,493 @@
...
+	TC_SUCCESS_RESULT();
+}
+
+static int wifi_on_callback(void *cb_data, void *event_data)
+{
+	static int cnt = 0;
+
+	printf("WIFI ON CALLBACK!! %d\n", ++cnt);
+
+	if (cb_data == NULL && event_data != NULL && strncmp(event_data, EL_WIFI_ON_DATA, sizeof(EL_WIFI_ON_DATA) + 1) == 0) {
```
Comment (jeongchanKim) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: for strncmp, sizeof(EL_WIFI_ON_DATA) + 1 is correct? not sizeof(EL_WIFI_ON_DATA) or strlen(EL_WIFI_ON_DATA) + 1?

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 17]
```diff
@@ -0,0 +1,493 @@
...
+	return EVENTLOOP_CALLBACK_STOP;
+}
+
+static void wifi_off_callback(void *cb_data, void *event_data)
+{
+	int ret;
+
+	printf("WIFI OFF CALLBACK!!\n");
+
+	if (cb_data != NULL && strncmp(cb_data, EL_CB_DATA, sizeof(EL_CB_DATA) + 1) == 0 && event_data != NULL && strncmp(event_data, EL_WIFI_OFF_DATA, sizeof(EL_WIFI_OFF_DATA) + 1) == 0) {
```
Comment (jeongchanKim) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: same as above

File: apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c [hunk 18]
```diff
@@ -0,0 +1,493 @@
...
+	g_repeat_timer = eventloop_add_timer(1000, true, repeat_timer_cb, EL_REPEAT_TIMER_DATA);
+	TC_ASSERT_NEQ("eventloop_add_timer", g_repeat_timer, NULL);
+	/* Add timer which will stop above repeated timer, g_repeat_timer after 5s. */
+	timer = eventloop_add_timer(5000, false, (timeout_callback)repeat_timer_stop_cb, EL_REPEAT_TIMER_DATA);
+	TC_ASSERT_NEQ("eventloop_add_timer", timer, NULL);
+	ret = eventloop_loop_run();
+	TC_ASSERT_EQ("eventloop_loop_run", ret, OK);
+	TC_ASSERT_EQ("eventloop_add_timer", el_timer_flag, 4);
+
+	/* Add timer which repeats forever, it is stooped by calling eventloop_loop_stop. */
```
Comment (jeongchanKim) [inline apps/examples/testcase/ta_tc/eventloop/utc/utc_eventloop_main.c]: Please check the spell. 
stooped -> stopped

**Comment by seinfra** [comment]
Target : [7caf44d78e20fa5985639ef8cf085fd8225aac40] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [7caf44d78e20fa5985639ef8cf085fd8225aac40] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [230d5f5854ab0c3ede5cb2b6059d32feb6f633dc] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [230d5f5854ab0c3ede5cb2b6059d32feb6f633dc] - Code Rule Check OK.

**Pull request merged**